clang/lib/CodeGen/Targets/AArch64.cpp

```diff
@@ -789,27 +791,37 @@ Address AArch64ABIInfo::EmitAAPCSVAArg(Address VAListAddr, QualType Ty,
                                  OnStackBlock, "vaargs.addr");
 
   if (IsIndirect)
-    return Address(CGF.Builder.CreateLoad(ResAddr, "vaarg.addr"), ElementTy,
-                   TyAlign);
-
-  return ResAddr;
+    return CGF.EmitLoadOfAnyValue(
+        CGF.MakeAddrLValue(
+            Address(CGF.Builder.CreateLoad(ResAddr, "vaarg.addr"), ElementTy,
+                    TyAlign),
+            Ty),
+        Slot);
+
+  return CGF.EmitLoadOfAnyValue(CGF.MakeAddrLValue(ResAddr, Ty), Slot);
 }
 
-Address AArch64ABIInfo::EmitDarwinVAArg(Address VAListAddr, QualType Ty,
-                                        CodeGenFunction &CGF) const {
+RValue AArch64ABIInfo::EmitDarwinVAArg(Address VAListAddr, QualType Ty,
+                                       CodeGenFunction &CGF,
+                                       AggValueSlot Slot) const {
   // The backend's lowering doesn't support va_arg for aggregates or
   // illegal vector types.  Lower VAArg here for these cases and use
   // the LLVM va_arg instruction for everything else.
   if (!isAggregateTypeForABI(Ty) && !isIllegalVectorType(Ty))
-    return EmitVAArgInstr(CGF, VAListAddr, Ty, ABIArgInfo::getDirect());
+    return CGF.EmitLoadOfAnyValue(
+        CGF.MakeAddrLValue(
+            EmitVAArgInstr(CGF, VAListAddr, Ty, ABIArgInfo::getDirect()), Ty),
+        Slot);
 
   uint64_t PointerSize = getTarget().getPointerWidth(LangAS::Default) / 8;
   CharUnits SlotSize = CharUnits::fromQuantity(PointerSize);
 
   // Empty records are ignored for parameter passing purposes.
-  if (isEmptyRecord(getContext(), Ty, true))
-    return Address(CGF.Builder.CreateLoad(VAListAddr, "ap.cur"),
-                   CGF.ConvertTypeForMem(Ty), SlotSize);
+  if (isEmptyRecord(getContext(), Ty, true)) {
+    Address ResAddr = Address(CGF.Builder.CreateLoad(VAListAddr, "ap.cur"),
```

Comment:
Maybe we should just skip copying empty records?

Reply:
Yeah, for this and the point above, it feels like we should be able to just do nothing. (`Ignore` can only be used for empty types.)

Reply:
Then we will end up with a temporary alloca but not using it. Would that be still ok?

Reply:
You mean, the alloca for the destination?  Sure, that's fine.

Reply:
It's fine to have an `alloca` that ends up unused.  It'd be slightly better not to, so if you can easily avoid the allocation by e.g. doing the empty-class check earlier, please do; but if that's awkward to achieve, don't worry about it.

Reply:
Done.

Reply:
This isn't returning the correct address; for aggregates, we should always return the address in "Slot".

Reply:
That said, it might not really matter in this context... CGExprAgg ignores the return value.  But still, there's no point to constructing a load that will never be used.

Reply:
Good point, I didn't think about that. I removed load construction.